datatypes/mcs_decimal.h

```diff
@@ -441,226 +442,93 @@ class Decimal: public TSInt128
             return (std::get<0>(integralFractionalDivisor) % div.getValue()) * std::get<2>(integralFractionalDivisor) + std::get<1>(integralFractionalDivisor);
         }
 
-        bool operator==(const Decimal& rhs) const
+        template<typename Op128, typename Op64>
+        bool cmpOperatorTemplate(const Decimal& rhs) const
         {
+            Op128 op128;
+            Op64 op64;
             if (precision > datatypes::INT64MAXPRECISION &&
                 rhs.precision > datatypes::INT64MAXPRECISION)
             {
                 if (scale == rhs.scale)
-                    return s128Value == rhs.s128Value;
+                    return op128(s128Value, rhs.s128Value);
                 else
-                    return (datatypes::Decimal::compare(*this, rhs) == 0);
+                    return op64(datatypes::Decimal::compare(*this, rhs), 0);
             }
             else if (precision > datatypes::INT64MAXPRECISION &&
                      rhs.precision <= datatypes::INT64MAXPRECISION)
             {
-                const_cast<Decimal&>(rhs).s128Value = rhs.value;
 
                 if (scale == rhs.scale)
-                    return s128Value == rhs.s128Value;
+                {
+                    return op128(s128Value, rhs.s128Value);
+                }
                 else
-                    return (datatypes::Decimal::compare(*this, rhs) == 0);
+                {
+                    //const_cast<Decimal&>(rhs).s128Value = rhs.value;
+                    return op64(datatypes::Decimal::compare(*this, 
+                                                            Decimal(scale,
+                                                                    precision,
+                                                                    (int128_t) rhs.value)),
+                                0);
```

Comment:
Should not it be `Decimal(rhs.scale, rhs.precision, (int128_t) rhs.value))` here?